sw/device/silicon_creator/lib/shutdown.c

```diff
@@ -269,12 +325,40 @@ SHUTDOWN_FUNC(NO_MODIFIERS, shutdown_report_error(rom_error_t reason)) {
   // that we won't jump to a different function.
   uint32_t redacted_error = shutdown_redact_inline(reason, policy);
 
-  // TODO(lowrisc/opentitan#7894): What (if anything) should we print at
-  // startup?
-  log_printf("lc_state: 0x%x\n\r", (unsigned int)raw_state);
-
-  // TODO(lowRISC/opentitan#8236): log_printf is in the .text section.
-  log_printf("boot_fault: 0x%x\n", (unsigned int)redacted_error);
+  // Reset UART TX fifo and enable TX.
+  abs_mmio_write32(kUartBase + UART_FIFO_CTRL_REG_OFFSET,
+                   bitfield_bit32_write(0, UART_FIFO_CTRL_TXRST_BIT, true));
+  uint32_t uart_ctrl_reg = abs_mmio_read32(kUartBase + UART_CTRL_REG_OFFSET);
+  uart_ctrl_reg = bitfield_bit32_write(uart_ctrl_reg, UART_CTRL_TX_BIT, true);
+  abs_mmio_write32(kUartBase + UART_CTRL_REG_OFFSET, uart_ctrl_reg);
+
+  // Print the error message.
+  const char kErrorPrefix[kMsgPrefixLen] = "boot_fault: 0x";
+  shutdown_print_prefix(kErrorPrefix);
+  shutdown_print32(redacted_error);
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, '\n');
+
+  // Wait until UART TX is idle.
+  bool is_empty = false;
+  do {
+    is_empty =
+        bitfield_bit32_read(abs_mmio_read32(kUartBase + UART_STATUS_REG_OFFSET),
+                            UART_STATUS_TXIDLE_BIT);
+  } while (!is_empty);
+
+  // Print the raw life cycle state as reported by the hardware.
+  const char kLcStatePrefix[kMsgPrefixLen] = "lc_state  : 0x";
+  shutdown_print_prefix(kLcStatePrefix);
+  shutdown_print32(raw_state);
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, '\n');
+
+  // Wait until UART TX is idle.
+  is_empty = false;
+  do {
+    is_empty =
+        bitfield_bit32_read(abs_mmio_read32(kUartBase + UART_STATUS_REG_OFFSET),
```

Comment:
Why not
```c
while(bitfield_bit32_read(...)) {}
```

Reply:
It actually looked worse than this with the abs_mmio_read32 but I split it so that the register is read in the loop body.